plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java

```diff
@@ -168,54 +159,144 @@ protected void canceling() {
+    String jsonCredential = new CredentialHelper().toJson(credential);
+    try {
+      Files.write(destination, jsonCredential.getBytes(StandardCharsets.UTF_8));
+    } catch (IOException ex) {
+      return StatusUtil.error(getClass(), Messages.getString("save.credential.failed"), ex);
     }
+    return Status.OK_STATUS;
   }
 
-  private static void saveCredential(Path destination, Credential credential) throws IOException {
-    String jsonCredential = new CredentialHelper().toJson(credential);
-    Files.write(destination, jsonCredential.getBytes(StandardCharsets.UTF_8));
+  private IStatus stageProject(Path credentialFile, IPath explodedWarDirectory, IPath stagingDirectory, IProgressMonitor monitor) {
+    SubMonitor progress = SubMonitor.convert(monitor, 100);
+    RecordProcessError stagingExitListener = new RecordProcessError();
+    CloudSdk cloudSdk = getCloudSdk(credentialFile, stagingStdoutLineListener, stagingExitListener);
+
+    try {
+      getJobManager().beginRule(project, progress);
+      new ExplodedWarPublisher().publish(project, explodedWarDirectory, progress.newChild(40));
+      new StandardProjectStaging().stage(explodedWarDirectory, stagingDirectory, cloudSdk, progress.newChild(60));
+      return stagingExitListener.getExitStatus();
+    } catch (CoreException | IllegalArgumentException | OperationCanceledException ex) {
+      return StatusUtil.error(getClass(), Messages.getString("deploy.job.staging.failed"), ex);
+    } finally {
+      getJobManager().endRule(project);
+    }
+  }
+
+  private IStatus deployProject(Path credentialFile, IPath stagingDirectory, IProgressMonitor monitor) {
+    RecordProcessError deployExitListener = new RecordProcessError();
+    CloudSdk cloudSdk = getCloudSdk(credentialFile, deployStdoutLineListener, deployExitListener);
+    new AppEngineProjectDeployer().deploy(
+        stagingDirectory, cloudSdk, deployConfiguration, monitor);
+    return deployExitListener.getExitStatus();
   }
 
-  private CloudSdk getCloudSdk(Path credentialFile) {
+  private CloudSdk getCloudSdk(Path credentialFile, ProcessOutputLineListener stdoutLineListener, ProcessExitListener processExitListener) {
     CloudSdk cloudSdk = new CloudSdk.Builder()
                           .addStdOutLineListener(stdoutLineListener)
                           .addStdErrLineListener(stderrLineListener)
                           .addStdErrLineListener(errorCollectingLineListener)
                           .appCommandCredentialFile(credentialFile.toFile())
                           .startListener(new StoreProcessObjectListener())
-                          .exitListener(new RecordProcessError())
+                          .exitListener(processExitListener)
                           .appCommandMetricsEnvironment(CloudToolsInfo.METRICS_NAME)
                           .appCommandMetricsEnvironmentVersion(CloudToolsInfo.getToolsVersion())
+                          .appCommandOutputFormat("json")
```

Comment:
Will this change the error output to JSON as well? If so, it might interfere with the error message parsing via `errorCollectingLineListener`.